packages/libs/client/src/v2/pact.ts

```diff
@@ -0,0 +1,238 @@
+/* eslint-disable @kadena-dev/no-eslint-disable */
+/* eslint-disable @typescript-eslint/explicit-function-return-type */
+
+import { createExp } from '@kadena/pactjs';
+
+import { parseType } from '../utils/parseType';
+
+const prop = <T extends any>(name: string, value?: T) =>
+  value === undefined ? {} : { [name]: value };
+
+export const getModule = (name: string) => {
+  let code = name;
+  const pr: any = new Proxy<any>(function () {} as unknown, {
+    get(_, path: string) {
+      code += '.' + path;
+      return pr;
+    },
+    apply(_, thisArg, args) {
+      const exp = createExp(code, ...args.map(parseType));
+      code = name;
+      return exp;
+    },
+  });
+  return pr;
+};
+
+interface IPayload {
+  exec: <
+    T extends Array<{
+      capability(name: string, ...args: any): ICapabilityItem;
+    }>,
+  >(
+    codes: T,
+    data?: any,
+    // use _branch to add type inferring for using it when user call signer function then we can show a related list of capabilities
+  ) => { payload: IExecPayload & { funs: [...T]; _brand: 'exec' } };
+  cont: (options: IContinuationPayload) => {
+    payload: IContinuationPayload & { _brand: 'cont' };
+  };
+}
+
+export const payload: IPayload = {
+  exec: (codes, data) => ({
+    payload: {
+      code: codes.join(''),
+      ...prop('data', data),
+    } as any,
+  }),
+  cont: (options) => ({
+    payload: options as any,
+  }),
+};
+
+export type IPartialCommand = Partial<Omit<ICommand, 'payload'>>;
+
+// TODO : improve the return value to merge all of the inputs as an object
+interface ICommandBuilder {
+  <F extends Pick<ICommand, 'payload'>>(
+    payload: F,
+    ...second: [
+      ...Array<
+        | Partial<ICommand>
+        | ((payload: F & Partial<ICommand>) => Partial<ICommand>)
+      >,
+    ]
+  ): { command: Partial<ICommand>; stringify: () => string };
+}
+
+export const commandBuilder: ICommandBuilder = (first: any, ...rest: any) => {
```

Comment:
I'd rather have the command builder become lazy, which would allow for further composition after initial composition.

Reply:
Yes its more flexible and I have tried this approach, but I couldn't find a way to write a proper typing for that
Actually the most difficult typing is capabilities in signer, that should refer types from the payload and only suggest the valid capabilities. So I had to add some limitations like forcing the first argument to be the payload. 
so basically 
```typescript
commandBuilder(
payload.exec([
  coin.transfer(from, to, { decimal: '0.01' })
]),
// withCapability comes you the proper typings 
 signer(from, (withCapability) => [
      withCapability('coin.TRANSFER', from, to, { decimal: amount }),
      withCapability('coin.TRANSFER', from, to, { decimal: '0.01' }),
    ]),
)
```

Reply:
I have removed the restriction, so now users can compose multiple commands easily. However, the command builder still returns the ICommand object. Although I didn't implement laziness, the input arguments can still be functions that receive the current state, such as the `signer`.

I have check the typing again, actually it would be technically straightforward to implement laziness. However, based on the current requirements, there doesn't seem to be a necessity for it since we have divided the signing and submitting steps.

Your example does not need lazyness as all of the helpers can just return the partial command and don't need to wait for any input.
```TypeScript
setMeta(devnetInfo) // probably just returns the {meta: ... }
setSigner(devnetAccount) // can return {signers:[...]}
getTransactionCreate(from, to, keysetAttributeInData, amount) // can return {payload:...}
setData(dataWithKeyset) // can set the {payload:data{...}}
```
The only concern I have about implementing laziness is that it would require all users to make two function calls in a row for typical use cases , and I'm unsure how often this is needed.
```
const command = commandBuilder(
 .....
)(/ send initial state/)
```
@alber70g any idea?

Reply:
In worse case you could introduce a `commandAndSend` that directly executes it right?
```
const commandAndSend = (
  .....
  );
  ```